be/src/exprs/math_functions.cpp

```diff
@@ -851,4 +851,113 @@ template StatusOr<ColumnPtr> MathFunctions::cosine_similarity<TYPE_FLOAT, true>(
+
+    if (base_flat_size != target_flat_size) {
+        return Status::InvalidArgument("l2_distance requires equal length arrays");
+    }
+
+    if (base_flat->has_null() || target_flat->has_null()) {
+        return Status::InvalidArgument("l2_distance does not support null values");
+    }
+    if (base_flat->is_nullable()) {
+        base_flat = down_cast<const NullableColumn*>(base_flat)->data_column().get();
+    }
+    if (target_flat->is_nullable()) {
+        target_flat = down_cast<const NullableColumn*>(target_flat)->data_column().get();
+    }
+
+    using CppType = RunTimeCppType<TYPE>;
+    using ColumnType = RunTimeColumnType<TYPE>;
+
+    const CppType* base_data_head = down_cast<const ColumnType*>(base_flat)->get_data().data();
+    const CppType* target_data_head = down_cast<const ColumnType*>(target_flat)->get_data().data();
+
+    // prepare result with nullable value.
+    ColumnPtr result = ColumnHelper::create_column(TypeDescriptor{TYPE}, false, false, target_size);
+    ColumnType* data_result = down_cast<ColumnType*>(result.get());
+    CppType* result_data = data_result->get_data().data();
+
+    for (size_t i = 0; i < target_size; i++) {
+        size_t t_dim_size = target_offset[i + 1] - target_offset[i];
+        size_t b_dim_size = base_offset[i + 1] - base_offset[i];
+        if (t_dim_size != b_dim_size) {
+            return Status::InvalidArgument(
+                    fmt::format("l2_distance requires equal length arrays in each row. base array dimension size "
+                                "is {}, target array dimension size is {}.",
+                                b_dim_size, t_dim_size));
+        }
+        if (t_dim_size == 0) {
+            return Status::InvalidArgument("l2_distance requires non-empty arrays in each row");
+        }
+    }
+
+    const CppType* target_data = target_data_head;
+    const CppType* base_data = base_data_head;
+
+    for (size_t i = 0; i < target_size; i++) {
+        CppType sum = 0;
+        CppType base_magnitude = 0;
+        CppType target_magnitude = 0;
+        size_t dim_size = target_offset[i + 1] - target_offset[i];
+        for (size_t j = 0; j < dim_size; j++) {
+            base_magnitude += base_data[j] * base_data[j];
+            target_magnitude += target_data[j] * target_data[j];
+        }
+        target_magnitude = std::sqrt(target_magnitude);
+        base_magnitude = std::sqrt(base_magnitude);
+        for (size_t j = 0; j < dim_size; j++) {
+            CppType distance;
+            if constexpr (isNorm) {
+                distance = (base_data[j] / base_magnitude - target_data[j] / target_magnitude)
+                        * (base_data[j] / base_magnitude - target_data[j] / target_magnitude);
+            } else {
+                distance = (base_data[j] - target_data[j]) * (base_data[j] - target_data[j]);
+            }
+            sum += distance;
+        }
+        result_data[i] = sum;
```

Comment:
If this is an l2 distance by definition, maybe we should sqrt again? 
`result_data[i] = std::sqrt(sum);`

Reply:
> The l2_distance used here refers to the distance without taking the square root, which is consistent with the behavior in faiss.
The consensus is usually calculated in this way. However, in our internal tests, after implementing vector indexing with SR + faiss, we found that the l2_distance calculated by faiss's index is not consistent with the result of the brute force algorithm. It turns out that faiss's l2_distance does not use sqrt. We are unsure if we need to align this on our end. I am fine with either adding it or not~